datafusion/optimizer/src/common_subexpr_eliminate.rs

```diff
@@ -351,56 +460,108 @@ impl CommonSubexprEliminate {
             &mut expr_stats,
             ExprMask::NormalAndAggregates,
         )?;
-        let mut common_exprs = IndexMap::new();
-        let mut rewritten_exprs = self.rewrite_exprs_list(
-            vec![new_aggr_expr.clone()],
-            &[&aggr_arrays],
-            &expr_stats,
-            &mut common_exprs,
-            &config.alias_generator(),
-        )?;
-        transformed |= rewritten_exprs.transformed;
-        let rewritten = pop_expr(&mut rewritten_exprs.data)?;
+        if aggr_found_common {
+            let mut common_exprs = CommonExprs::new();
+            let mut rewritten_exprs = self.rewrite_exprs_list(
+                vec![new_aggr_expr.clone()],
+                vec![aggr_arrays],
+                &expr_stats,
+                &mut common_exprs,
+                &config.alias_generator(),
+            )?;
+            assert!(rewritten_exprs.transformed);
+            let rewritten = pop_expr(&mut rewritten_exprs.data)?;
+
+            assert!(!common_exprs.is_empty());
+            let mut agg_exprs = common_exprs
+                .into_values()
+                .map(|(expr, expr_alias)| expr.alias(expr_alias))
+                .collect::<Vec<_>>();
+
+            let new_input_schema = Arc::clone(new_input.schema());
+            let mut proj_exprs = vec![];
+            for expr in &new_group_expr {
+                extract_expressions(expr, &new_input_schema, &mut proj_exprs)?
+            }
+            for (expr_rewritten, expr_orig) in rewritten.into_iter().zip(new_aggr_expr) {
+                if expr_rewritten == expr_orig {
+                    if let Expr::Alias(Alias { expr, name, .. }) = expr_rewritten {
+                        agg_exprs.push(expr.alias(&name));
+                        proj_exprs.push(Expr::Column(Column::from_name(name)));
+                    } else {
+                        let expr_alias = config.alias_generator().next(CSE_PREFIX);
+                        let (qualifier, field) =
+                            expr_rewritten.to_field(&new_input_schema)?;
+                        let out_name = qualified_name(qualifier.as_ref(), field.name());
+
+                        agg_exprs.push(expr_rewritten.alias(&expr_alias));
+                        proj_exprs.push(
+                            Expr::Column(Column::from_name(expr_alias)).alias(out_name),
+                        );
+                    }
+                } else {
+                    proj_exprs.push(expr_rewritten);
+                }
+            }
+
+            let agg = LogicalPlan::Aggregate(Aggregate::try_new(
+                new_input,
+                new_group_expr,
+                agg_exprs,
+            )?);
 
-        if common_exprs.is_empty() {
+            Projection::try_new(proj_exprs, Arc::new(agg))
+                .map(LogicalPlan::Projection)
+                .map(Transformed::yes)
+        } else {
+            // TODO: How exactly can the name or the schema change in this case?
```

Comment:
Is this something you plan to do in this PR? Or is it for follow up work (I can file a new ticket if it is follow on work)

Reply:
Yes. I didn't want to remove this part in this PR, but I think we can get rid of it in a follow-up one. If you file a ticket then please ping me or assign it to me. I will be offline for a about a week but when I'm back I'm happy to fix this.